### ISSUE TITLE
docker dlv headless

### ISSUE THREAD

**Issue by gertcuykens** [microsoft/vscode-go#1806]
vscode-go 0.6.84

Step 1 make app
```go
func main() {
	v := "test"
	fmt.Println(v)
	fmt.Println("Hello World!")
}
```
Step 2 create vscode debug config (PS: some string magic would be nice so you don't have to hardcode your remotePath ${remove /Users/gert from filedir path magic})
```json
  "version": "0.2.0",
    "configurations": [
        {
            "name": "docker",
            "type": "go",
            "request": "launch",
            "mode": "remote",
            "backend": "native",
            "remotePath": "/go/src/myapp",
            "port": 2345,
            "host": "127.0.0.1",
            "program": "${fileDirname}",
            "env": {},
            "args": [],
            "showLog": true
        }
```
Step 3 create docker container and start it up in bash
```sh
docker run -it --rm \
                -v /Users/gert/go:/go \
                -w /go/src/myapp \
                -p 2345:2345 \
                --security-opt apparmor=unconfined \
                --cap-add SYS_PTRACE \
                --privileged \
                myappcontainer bash
```
Step 4 in your container launch headless dlv
```
dlv debug --headless --listen=:2345 --log
```

Step 5 breakpoints work until you try to step into everything that is inside GOROOT

![image](https://user-images.githubusercontent.com/863731/43160002-3a99e868-8f84-11e8-9d45-4770fbad0a4f.png)

Step 6 and last but not least, the restart button is not working

![image](https://user-images.githubusercontent.com/863731/43160239-e8961090-8f84-11e8-8ebb-1d95b997be3e.png)

```
DEBU[0051] continuing                                    layer=debugger
test
DEBU[0069] stepping                                      layer=debugger
DEBU[0341] halting                                       layer=debugger
[1]+  Stopped  
```
Should not halt just restart the headless server probably related to https://github.com/derekparker/delve/issues/1279

**Comment by mmwtsn** [comment]
Seeing this as well.

Just curious, why did you add the security option and capacity add flags in your example? I haven't used those before and was able to reproduce the issue without either of them here.

**Comment by ramya-rao-a** [comment]
@gertcuykens @mmwtsn If you still see this with the latest version of the extension, do you mind running the extension from a local setup and debugging it? It is pretty straight-forward

- Follow the steps in [Building & Debugging the Go extension](https://github.com/Microsoft/vscode-go/wiki/Building,-Debugging-and-Sideloading-the-extension-in-Visual-Studio-Code#building-and-debugging-the-extension) to get a local set up of the extension running
- Follow the steps in [Debugging the debugger](https://github.com/Microsoft/vscode-go/tree/master/src/debugAdapter), add a breakpoint at https://github.com/Microsoft/vscode-go/blob/0.6.87/src/debugAdapter/goDebug.ts#L654

That is where the remote path gets translated to a local path, and thats where the translation goes bad and you end up with paths like `/usr/local/Users/gert/go/src/fmt/print.go` which of course doesnt exist

**Comment by gertcuykens** [comment]
yes it's still the same on 0.6.88 wait i give you a dockerfile to reproduce it

```
FROM golang:rc-alpine as root
WORKDIR /go/src/yourapp
RUN apk add --no-cache ca-certificates musl-dev gcc file git vim curl \
    && apk update \
    && apk upgrade \
    && rm -rf /var/cache/apk/*
CMD ["ash"]

FROM root as dev
COPY . .
RUN go get -u github.com/golang/dep/cmd/dep
RUN go get -u github.com/derekparker/delve/cmd/dlv
RUN dep ensure
RUN go build -a -tags 'netgo noplugin' -ldflags '-d -s -w -extldflags "-static"' -o /app/bin/main
RUN go build -a -tags 'netgo noplugin' -ldflags '-d -extldflags "-static"' -gcflags "all=-N -l" -o /app/bin/debug
RUN ldd /app/bin/main; exit 0
RUN ldd /app/bin/debug; exit 0
RUN file /app/bin/main
RUN file /app/bin/debug
# CMD ["/app/bin/main"]
CMD ["dlv", "--listen=:2345", "--headless", "--api-version=2", "--accept-multiclient", "exec", "/app/bin/debug"]
EXPOSE 2345
```
```
docker build -t yourapp . --target dev
docker run -it --rm --privileged -p 2345:2345 yourapp
```
Now before you start running vscode we need to test dlv cli first, don't skip this because the remote config shenanigans from vscode are confusing because dlv has no `remotePath` in the first place. It's only has 2 things a 
```
--wd string            Working directory for running the program. (default ".") 
which in my opinion never has to be changed ever and doesn't concern vscode dlv client at all 
just the headless server
```
 and a `substitute-path` in the dlv client
```
(dlv) config -list
aliases            map[]
substitute-path    [{/go /Users/gert/go}]
max-string-len     <not defined>
max-array-values   <not defined>
show-location-expr false
(dlv)
```
note that headless has a complete new way to restart a headless session
```
(dlv) r
Process restarted with PID 25
(dlv) q
Would you like to kill the headless instance? [Y/n]
```
Now how importent is this to make this work, lets say this is the only solution to debug a kubernetes pod :P And microsoft loves kubernetes like the rest of the world

My suggestion is tackle the vscode config itself first and make it inline with dlv and less confusing  

```
  "configurations": [
        {
            "name": "docker",
            "type": "go",
            "request": "launch",
            "mode": "remote",
            "backend": "native",
            "substitute-path": [{/go /Users/gert/go}] (do config thing in dlv before debug starts),
            "port": 2345,
            "host": "127.0.0.1",
            "working-directorie": "${fileDirname}", (wd dlv)
            "env": {},
            "args": [], make sure this do -- ... see dlv cli instructions
            "showLog": true
        }
```
So for me its important that your dlv cli on your macbook is doing the same thing as on my macbook before we start figuring out the source code. Because I believe it needs a bigger refactor then anticipated.

**Comment by gertcuykens** [comment]
Was thinking a bit more about this what if we just remove `"remotePath": "/go/src/myapp",` and `"program": "${fileDirname}"` completly for mode remote and concentrate more on
```  
  "go.delveConfig": {
        "useApiV1": false,
        "dlvLoadConfig": {
            "followPointers": true,
            "maxVariableRecurse": 1,
            "maxStringLen": 64,
            "maxArrayValues": 64,
            "maxStructFields": -1
        }
    }
```
Suggest to replace `"dlvLoadConfig":{}` to "dlvLoadConfig":"~/.dlv/config.yml" so we are using the exact same config file as the dlv client and not introduce another configuration space that does exactly the same thing and creates confusion as in dlv working and vscode not because each using separate config.

**Comment by ramya-rao-a** [comment]
@gertcuykens I got the below error when building a docker image using your docker file

<img width="611" alt="screen shot 2018-08-23 at 6 09 04 pm" src="https://user-images.githubusercontent.com/16890566/44559568-d9751380-a6ff-11e8-9db0-46f5f9ee1fe3.png">

> the remote config shenanigans from vscode are confusing because dlv has no remotePath in the first place.

The `remotePath` in the VS Code debug config is not passed to delve. It is solely used to map the file paths in the remote system (used by delve headless server) to the file paths in the local box. So I don't see how remote debugging can work without it.

The particular issue you are seeing where breakpoints work until you step into std library code is definitely due to a bug in mapping the path sent by delve headless server to the local path. So if you can please follow the instructions in 
https://github.com/Microsoft/vscode-go/issues/1806#issuecomment-413393069, I am sure we can catch the bug.

**Comment by ramya-rao-a** [comment]
I made the below changes to your docker file
- set `WORKDIR` to `app` instead of `yourapp`
- removed `dep` and `ldd` related entries
- built the image with `docker build -t app . --target dev`
- ran the container with `docker run -it --rm --privileged -p 2345:2345 app`

Below is my code set-up
- A folder called `app` with file `main.go` with your code sample
- The above folder is opened in vscode with the below debug config in the launch.json file
```
{
            "name": "Launch",
            "type": "go",
            "request": "launch",
            "mode": "remote",
            "remotePath": "/go/src/app",
            "port": 2345,
            "host": "127.0.0.1",
            "program": "${fileDirname}",
            "env": {},
            "args": [],
            "showLog": true,
            "trace": "verbose",
            "useApiV1": false
        }
```

Once the container was up and running, I added a breakpoint in VS Code and then pressed F5 to start debugging. The breakpoint was successfully hit.

In the interactive container, I could see `Hello World` after continuing from the breakpoint.

The second attempt, resulted in failure in creating the breakpoint and the error `Process 15 has exited with status 0`. 

I believe the main issue here is the inability to restart debugging the program once it has exited i.e a VS Code counterpart for the `r` command you use in the dlv client.

I read up on the [substitute-path configuration](https://github.com/derekparker/delve/blob/9a216211d3461ab031f822c012bc27cab9758da0/pkg/config/config.go#L136-L142). Seems like it is only useful when sources are moved to a different place
between compilation and debugging. That is not the case here. So I don't believe adding support for that would change anything.

**Comment by gertcuykens** [comment]
A great now you can step into a GOROOT function to see if your changes fixed the GOROOT bug.
So to recap 
- Only dlv cli needs a `substitute-path` or it's not able to figure out the correct path
- Like you noticed the headless server needs a complete new way to 
 a) restart a process
 b) shutdown a process 
```
$ dlv connect localhost:2345
Type 'help' for list of commands.
(dlv) q
Would you like to kill the headless instance? [Y/n]
```
But lets do the restart part in a separate ticket and focus on remote path only, so if you are successful stepping into `print.go` and your code is located in your docker container on a different place example `/Users/gert/go` vs `/go` then you solved it. 

Note that for dlv cli the source code is moved from /Users/gert/go to /go in the docker. As you will see the dlv cli will not work if you don't add a substitute-path.

**Comment by ramya-rao-a** [comment]
I am in transit today, will check the step in to GOROOT later.
Are you coming to Gophercon this week? If so, we can look into this together then.

**Comment by gertcuykens** [comment]
Now you made me even more want to go to Gopercon... but I can't I am stuck in Belgium I need to develop some golang app and deploy it into a kubernetes cluster or I will dispoint my colleagues. Though I will be watching your talk if you are giving one :D

**Comment by quoctruong** [comment]
@gertcuykens @ramya-rao-a I recently tested stepping into the `fmv.Println` function on the latest version of the Go Debugger and this seems to work.

**Comment by ramya-rao-a** [comment]
Based on @quoctruong's assessment above, the issue seems to have been resolved.
Since we haven't heard back from anyone else, I am closing this issue.

Thanks for reporting & Happy Coding!
